diff -Nur reiserfsprogs-3.6.6/include/swab.h reiserfsprogs-3.6.6-new/include/swab.h
--- reiserfsprogs-3.6.6/include/swab.h	2003-04-22 21:56:50.000000000 +0800
+++ reiserfsprogs-3.6.6-new/include/swab.h	2012-02-01 13:47:29.681658000 +0800
@@ -1,22 +1,67 @@
 /*
- *  Copyright 2002-2003 by Hans Reiser, licensing governed by reiserfs/README
+ *  Copyright 2002-2004 by Hans Reiser, licensing governed by 
+ *  reiserfsprogs/README
  */
-#ifndef _REISERFS_SWAB_H_
-#define _REISERFS_SWAB_H_
 
-#include <endian.h>
+#ifndef REISERFSPROGS_SWAB_H
+#define REISERFSPROGS_SWAB_H
 
-#if __BYTE_ORDER == __BIG_ENDIAN
+#include <endian.h>
+#include <asm/types.h>
 
-#define extern static
-#define __BYTEORDER_HAS_U64__
-#include <linux/byteorder/swab.h>
-#undef extern
+#define __swab16(x) \
+({ \
+        __u16 __x = (x); \
+        ((__u16)( \
+                (((__u16)(__x) & (__u16)0x00ffU) << 8) | \
+                (((__u16)(__x) & (__u16)0xff00U) >> 8) )); \
+})
+
+#define __swab32(x) \
+({ \
+        __u32 __x = (x); \
+        ((__u32)( \
+                (((__u32)(__x) & (__u32)0x000000ffUL) << 24) | \
+                (((__u32)(__x) & (__u32)0x0000ff00UL) <<  8) | \
+                (((__u32)(__x) & (__u32)0x00ff0000UL) >>  8) | \
+                (((__u32)(__x) & (__u32)0xff000000UL) >> 24) )); \
+})
+
+#define __swab64(x) \
+({ \
+        __u64 __x = (x); \
+        ((__u64)( \
+                (__u64)(((__u64)(__x) & (__u64)0x00000000000000ffULL) << 56) | \
+                (__u64)(((__u64)(__x) & (__u64)0x000000000000ff00ULL) << 40) | \
+                (__u64)(((__u64)(__x) & (__u64)0x0000000000ff0000ULL) << 24) | \
+                (__u64)(((__u64)(__x) & (__u64)0x00000000ff000000ULL) <<  8) | \
+                (__u64)(((__u64)(__x) & (__u64)0x000000ff00000000ULL) >>  8) | \
+                (__u64)(((__u64)(__x) & (__u64)0x0000ff0000000000ULL) >> 24) | \
+                (__u64)(((__u64)(__x) & (__u64)0x00ff000000000000ULL) >> 40) | \
+                (__u64)(((__u64)(__x) & (__u64)0xff00000000000000ULL) >> 56) )); \
+})
+
+
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+
+# define cpu_to_le16(val)                 (val)
+# define le16_to_cpu(val)                 (val)
+# define cpu_to_le32(val)                 (val)
+# define le32_to_cpu(val)                 (val)
+# define cpu_to_le64(val)                 (val)
+# define le64_to_cpu(val)                 (val)
+
+#elif __BYTE_ORDER == __BIG_ENDIAN
+
+# define cpu_to_le16(val)                 __swab16(val)
+# define le16_to_cpu(val)                 __swab16(val)
+# define cpu_to_le32(val)                 __swab32(val)
+# define le32_to_cpu(val)                 __swab32(val)
+# define cpu_to_le64(val)                 __swab64(val)
+# define le64_to_cpu(val)                 __swab64(val)
 
 #else
-
-#include <linux/byteorder/swab.h>
-
+# error "nuxi/pdp-endian archs are not supported"
 #endif
 
-#endif /* _REISERFS_SWAB_H_ */
+#endif /* REISERFS_SWAB_H */
diff -Nur reiserfsprogs-3.6.6/lib/misc.c reiserfsprogs-3.6.6-new/lib/misc.c
--- reiserfsprogs-3.6.6/lib/misc.c	2003-04-22 21:56:50.000000000 +0800
+++ reiserfsprogs-3.6.6-new/lib/misc.c	2012-02-01 13:46:58.262533000 +0800
@@ -187,9 +187,14 @@
 
 #include <unistd.h>
 #include <linux/unistd.h>
+#include <linux/version.h>
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,24)
 #define __NR_bad_stat64 195
 _syscall2(long, bad_stat64, char *, filename, struct stat64 *, statbuf);
+#else
+#define bad_stat64 stat64
+#endif
 
 #else
 
