diff -Naur a/setup.py b/setup.py
--- a/setup.py	2014-06-29 19:05:48.000000000 -0700
+++ b/setup.py	2015-02-05 13:00:08.671693950 -0800
@@ -787,6 +787,7 @@
                                libraries=math_libs) )
         # Detect SSL support for the socket module (via _ssl)
         search_for_ssl_incs_in = [
+                              '/usr/include/'
                               '/usr/local/ssl/include',
                               '/usr/contrib/ssl/include/'
                              ]
@@ -799,11 +800,12 @@
             if krb5_h:
                 ssl_incs += krb5_h
         ssl_libs = find_library_file(self.compiler, 'ssl',lib_dirs,
-                                     ['/usr/local/ssl/lib',
+                                     ['/usr/lib/ssl',
+                                      '/usr/local/ssl/lib',
                                       '/usr/contrib/ssl/lib/'
                                      ] )
 
-        if (ssl_incs is not None and
+        if (ssl_incs is not None or
             ssl_libs is not None):
             exts.append( Extension('_ssl', ['_ssl.c'],
                                    include_dirs = ssl_incs,
@@ -1389,8 +1391,9 @@
         #
         # You can upgrade zlib to version 1.1.4 yourself by going to
         # http://www.gzip.org/zlib/
-        zlib_inc = find_file('zlib.h', [], inc_dirs)
+        zlib_inc = find_file('zlib.h', [], inc_dirs+['/usr/include'])
         have_zlib = False
+	print("zlib_inc ===> %s" % zlib_inc)
         if zlib_inc is not None:
             zlib_h = zlib_inc[0] + '/zlib.h'
             version = '"0.0.0"'
@@ -1405,8 +1408,10 @@
                 if line.startswith('#define ZLIB_VERSION'):
                     version = line.split()[2]
                     break
+	    print("version ===> %s" % version)
             if version >= version_req:
                 if (self.compiler.find_library_file(lib_dirs, 'z')):
+		    print("host_platform ===> %s" % host_platform)
                     if host_platform == "darwin":
                         zlib_extra_link_args = ('-Wl,-search_paths_first',)
                     else:
